backend/copr_backend/pulp.py

```diff
@@ -195,27 +321,119 @@ def create_content(self, path, labels):
+        return False
+
+    def try_lock(self, repository, batch):
+        """
+        Periodically try to acquire the lock, and execute the modify_repository_content_locked() method.
+        """
+
+        while True:
+
+            # We don't have fair locking (locks-first => processes-first).  So to
+            # avoid potential indefinite waiting (see issue #1423) we check if the
+            # task isn't already processed _without_ having the lock.
+
+            if batch.check_processed(delete_if_not=False):
+                self.log.info("Task processed by other process (no-lock)")
+                return
+
+            try:
+                lockdir = os.environ.get(
+                    "COPR_TESTSUITE_LOCKPATH", "/var/lock/copr-backend")
+                with lock(repository, lockdir=lockdir, timeout=5, log=self.log):
+                    if self.modify_repository_content_locked(repository, batch):
+                        # While we still hold the lock, notify others we processed their
+                        # task.  Note that we do not commit in case of random exceptions
+                        # above.
+                        batch.commit()
+                        self.log.debug("Repository version and Publication created by this process")
+                        break
+
+            except LockTimeout:
+                continue  # Try again...
+
+            # we never loop, only upon timeout
+            assert False
+
+    def modify_repository_content(self, repository, add_content_units, remove_content_units):
+        """
+        Add and/or remove a list of artifacts to/from a repository
         https://pulpproject.org/pulp_rpm/restapi/#tag/Repositories:-Rpm/operation/repositories_rpm_rpm_modify
         """
         path = os.path.join(repository, "modify/")
         url = self.config["base_url"] + path
-        data = {"add_content_units": artifacts}
-        self.log.info("Pulp: add_content: %s %s", path, artifacts)
+        data = {"add_content_units": add_content_units or [], "remove_content_units": remove_content_units or []}
         return requests.post(url, json=data, **self.request_params)
 
+    def add_content(self, repository, artifacts):
+        """
+        Add a list of artifacts to a repository
+        https://pulpproject.org/pulp_rpm/restapi/#tag/Repositories:-Rpm/operation/repositories_rpm_rpm_modify
+        """
+        batch = BatchedAddRemoveContent(repository, artifacts, None, self.opts, self.log)
+        # Put our task to Redis DB and allow _others_ to process our
+        # own task.  This needs to be run _before_ the lock() call.
+        batch.make_request()
+        try:
+            self.try_lock(repository, batch)
+            self.log.info("Pulp: add_content: %s (%s)", repository, artifacts)
+            return True
+        except Exception as e:
+            self.log.info("Pulp: Failed to add content: %s %s %s", repository, artifacts, e)
```

Comment:
The log goes to `backend.log` (visible by user), could we by any chance leak the PULP key here through str(e)?